File: src-tauri/src/pattern/print.rs
use borsh::{BorshDeserialize, BorshSerialize};
use ordered_float::NotNan;

#[derive(Debug, Clone, PartialEq, BorshSerialize, BorshDeserialize)]
pub struct PrintSettings {
  pub font: Font,
  pub header: String,
  pub footer: String,
  pub margins: PageMargins,
  pub show_page_numbers: bool,
  pub show_adjacent_page_numbers: bool,
  pub center_chart_on_pages: bool,
}

impl Default for PrintSettings {
  fn default() -> Self {
    Self {
      font: Font::default(),
      header: String::new(),
      footer: String::new(),
      margins: PageMargins::default(),
      show_page_numbers: true,
      show_adjacent_page_numbers: true,
      center_chart_on_pages: true,
    }
  }
}

#[derive(Debug, Clone, PartialEq, BorshSerialize, BorshDeserialize)]
pub struct Font {
  pub name: String,
  pub size: u16,
  pub weight: u16,
  pub italic: bool,
}

impl Default for Font {
  fn default() -> Self {
    Self {
      name: String::from("Arial"),
      size: 12,
      weight: 400,
      italic: false,
    }
  }
}

pub type Inches = NotNan<f32>;

#[derive(Debug, Clone, PartialEq, BorshSerialize, BorshDeserialize)]
pub struct PageMargins {
  pub left: Inches,
  pub right: Inches,
  pub top: Inches,
  pub bottom: Inches,
  pub header: Inches,
  pub footer: Inches,
}

impl Default for PageMargins {
  fn default() -> Self {
    Self {
      left: NotNan::new(0.5).unwrap(),
      right: NotNan::new(0.5).unwrap(),
      top: NotNan::new(0.5).unwrap(),
      bottom: NotNan::new(0.5).unwrap(),
      header: NotNan::new(0.5).unwrap(),
      footer: NotNan::new(0.5).unwrap(),
    }
  }
}

File: src/schemas/pattern/print.ts
import { field } from "@dao-xyz/borsh";
export class Font {
  @field({ type: "string" })
  name: string;

  @field({ type: "u16" })
  size: number;

  @field({ type: "u16" })
  weight: number;

  @field({ type: "bool" })
  italic: boolean;

  constructor(data: Font) {
    this.name = data.name;
    this.size = data.size;
    this.weight = data.weight;
    this.italic = data.italic;
  }
}

export class PageMargins {
  @field({ type: "f32" })
  left: number;

  @field({ type: "f32" })
  right: number;

  @field({ type: "f32" })
  top: number;

  @field({ type: "f32" })
  bottom: number;

  @field({ type: "f32" })
  header: number;

  @field({ type: "f32" })
  footer: number;

  constructor(data: PageMargins) {
    this.left = data.left;
    this.right = data.right;
    this.top = data.top;
    this.bottom = data.bottom;
    this.header = data.header;
    this.footer = data.footer;
  }
}

export class PrintSettings {
  @field({ type: Font })
  font: Font;

  @field({ type: "string" })
  header: string;

  @field({ type: "string" })
  footer: string;

  @field({ type: PageMargins })
  margins: PageMargins;

  @field({ type: "bool" })
  showPageNumbers: boolean;

  @field({ type: "bool" })
  showAdjacentPageNumbers: boolean;

  @field({ type: "bool" })
  centerChartOnPages: boolean;

  constructor(data: PrintSettings) {
    this.font = data.font;
    this.header = data.header;
    this.footer = data.footer;
    this.margins = data.margins;
    this.showPageNumbers = data.showPageNumbers;
    this.showAdjacentPageNumbers = data.showAdjacentPageNumbers;
    this.centerChartOnPages = data.centerChartOnPages;
  }
}
